waspc/data/Generator/templates/server/src/auth/utils.ts

```diff
@@ -102,3 +187,64 @@ export async function validateAndGetAdditionalFields(data: {
   }
   return result;
 }
+
+export type EmailProviderData = {
+  password: string;
+  isEmailVerified: boolean;
+  emailVerificationSentAt: Date | null;
+  passwordResetSentAt: Date | null;
+}
+
+export type UsernameProviderData = {
+  password: string;
+}
+
+export type OAuthProviderData = {}
+
+export type ProviderData = {
```

Comment:
When we are serialising and deserialising data, we use this type as a helper to determine the expected structure of the `providerData`.

Calling `deserializeProviderData<'email'>(data)` will make the `data` of type `EmailProviderData` and type check it properly.

`providerName` is a `string` so it can't help us to narrow down the type.

Reply:
That makes sense, I still wonder if this is the best way to do in TS, are there any alternative ways to go about it? How do you feel about it, happy with it?

ALso, you didn't answer my question about naming.

Reply:
I'm okay with the current type setup for the provider data.

As for the name, it also sounds okay to me 🤷 

We have:
- provider name
- provider user id
- provider data <---- what ever data that's needed for that provider

Reply:
Ok! So only naming remains.

Problem with naming here is that "provider data" is already a concept we have -> it is that JSON data saved in the DB. But now you use the same term for a structure that maps from provider name to shape of provider data for that provider.

So let's do typical exercise in naming: what is this structure really?
It is as I said above:
> a structure that maps from provider name to shape of provider data for that provider

Also, it is not used in many places, nor is it user-facing, so a bit longer name shouldn't be an issue.

So let's try a naive name: `ProviderNameToData`. Or `ProviderDataByName`. Yeah, I think that is good. I would also write above it a short comment that this is helper type to be used for type level programming. Because otherwise, if you look at it as normal type, it can be hard to figure out why would it have this structure, what is this useful for.

Reply:
I've refactored it a bit in the meantime, now it's `PossibleProviderData` and it looks like this:

```ts
export type PossibleProviderData = {
  email: EmailProviderData;
  username: UsernameProviderData;
  google: OAuthProviderData;
  github: OAuthProviderData;
}
```

Reply:
Ok, I would say it is still not super clear from the name what it does, but at least it is clear it is not just "provider data" but something special, and that is an improvement.
I would maybe throw a comment above saying we have this type just for other type shenanigans and call it a day.